pyabc/epsilon/temperature.py

```diff
@@ -0,0 +1,502 @@
+                     acceptor_config)
+
+    def _update(self,
+                t: int,
+                get_weighted_distances: Callable[[], pd.DataFrame],
+                acceptance_rate: float,
+                acceptor_config):
+        """
+        Compute the temperature for time `t`.
+        """
+        # scheme arguments
+        kwargs = dict(
+            t=t,
+            get_weighted_distances=get_weighted_distances,
+            max_nr_populations=self.max_nr_populations,
+            pdf_norm=acceptor_config['pdf_norm'],
+            kernel_scale=acceptor_config['kernel_scale'],
+            prev_temperature=self.temperatures.get(t-1, None),
+            acceptance_rate=acceptance_rate,
+        )
+
+        if t >= self.max_nr_populations - 1:
+            # t is last time
+            temperature = 1.0
+        elif not self.temperatures and self.initial_temperature is not None:
+            if callable(self.initial_temperature):
+                # execute scheme
+                temperature = self.initial_temperature(**kwargs)
+            else:
+                # should be float value
+                temperature = self.initial_temperature
+        else:
+            # evalute schemes
+            temps = []
+            for scheme in self.schemes:
+                temp = scheme(**kwargs)
+                temps.append(temp)
+            logger.debug(f"Proposed temperatures: {temps}.")
+
+            # compute next temperature based on proposals and fallback
+            # should not be higher than before
+            fallback = self.temperatures[t - 1] \
+                if t-1 in self.temperatures else np.inf
+            proposed_value = self.aggregate_fun(temps)
+            # also a value lower than 1.0 does not make sense
+            temperature = max(min(proposed_value, fallback), 1.0)
+
+        # record found value
+        self.temperatures[t] = temperature
+
+    def __call__(self,
+                 t: int) -> float:
+        return self.temperatures[t]
+
+
+class TemperatureScheme:
+    """
+    A TemperatureScheme suggests the next temperature value. It is used as
+    one of potentially multiple schemes employed in the Temperature class.
+    This class is abstract.
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self,
+                 t: int,
+                 get_weighted_distances: Callable,
+                 max_nr_populations: int,
+                 pdf_norm: float,
+                 kernel_scale: str,
+                 prev_temperature: float,
+                 acceptance_rate: float):
+        pass
+
+
+class AcceptanceRateScheme(TemperatureScheme):
+    """
+    Try to keep the acceptance rate constant at a value of
+    `target_rate`. Note that this scheme will fail to
+    reduce the temperature sufficiently in later iterations, if the
+    problem's inherent acceptance rate is lower, but it has been
+    observed to give big feasible temperature leaps in early iterations.
```

Comment:
" it has been observed to give ... " -> can you add some examples/citation ?

Reply:
Well ... it's difficult to explain in detail here (I just saw this in various instances), but I added a sentence about its usage for initial temperature proposition.

Reply:
oki, otherwise seemed a bit "you have to trust us"